### ISSUE TITLE
Higher level function for lapack pstrf and pstf2

### ISSUE THREAD

**Issue by jordan-melendez** [scipy/scipy#10832]
Now that https://github.com/scipy/scipy/pull/10537 added `?pstrf` and `pstf2`, I am interested in starting discussion on potential higher level functions. It appears that this question was left open ended on that PR, and I haven't found the discussion continued elsewhere. I guess the basic question would be:
* Would pivoting be an option to existing functions, something like `cholesky(..., pivot=True)`, or entirely new functions? (This might also apply to how higher level functions for https://github.com/scipy/scipy/issues/10788 are ultimately written as well)
* For me, at least, it would be helpful to have this in for the 1.4 release :) is there still enough time to get this in? I'd be happy to contribute.

**Comment by mdhaber** [comment]
For `lu_factor`, I wanted to add a `complete_pivoting` parameter (default `False`). When `True`, it would use `?getc2` and return a tuple with three elements (the usual plus column permutation indices). `lu_solve` could simply check the number of elements in the input tuple to decide which LAPACK routine to call, so it wouldn't need an extra argument.

**Comment by jordan-melendez** [comment]
Makes sense. If that ends up happening, then it might make sense to coordinate the naming of this `complete_pivoting` parameter with `cho_factor`, and make `cho_solve` mirror `lu_solve`.

**Comment by ilayn** [comment]
I fully acknowledge the need for this. However I am more inclined to get rid of the `?_factor` and `?_solve` stuff and actually make this more streamlined inside `lu` and `cholesky` by somehow returning the factorization if needed and/or using Cython and accepting nD arrays for the left hand side. I don't have a concrete plan yet though and it requires some API design touching some backward compatibility pain points.

**Comment by ilayn** [comment]
My reasoning is that `?_solve` already accepts the low-level LAPACK representation and hence you can directly call the original LAPACK routine with what `?_factor` returns. Hence the user actually saves time by skipping these thin wrappers and directly calling the routines themselves via `scipy.linalg.lapack` namespace.

**Comment by mdhaber** [comment]
Certainly it is faster to skip `?_solve` and `?_factor` but they are a bit friendlier, if only because of the names, than the LAPACK wrappers themselves. We can certainly add this functionality to `lu` and `cholesky`, but why not also add it to `?_solve` and `?_factor` (unless you intend to deprecate them)?

**Comment by jordan-melendez** [comment]
@ilayn does this mean that `?_solve` and `?_factor` will be deprecated soon? I do slightly prefer `lu` and `cholesky`, and I would definitely be in favor of them accepting nD arrays. I usually find myself using numpy's `cholesky` for this very reason, but if scipy had both a `pivot` option and could broadcast over dimensions like numpy, then scipy would be the clear winner IMO.

**Comment by ilayn** [comment]
I wouldn't say they are going to be deprecated because that's quite often used out in the wild and would be quite an annoyance for not much of a return. Especially not without offering a better alternative.

Instead we can try to find a way to inject more functionality to the regular `lu`, `qr`, `cholesky` without breaking anything. But designing a better API is not so straightforward and needs some discussion on the list I guess. nD array inclusion is something I am also longing for a lot for `linalg.solve` but unfortunately I am sparing much less time than I wished these days.
